tests/api_test.py

```diff
@@ -2949,6 +2949,20 @@ def loss(A, x):
     with jax.checking_leaks():
       _ = jax.grad(loss)(A, x)  # doesn't crash
 
+  def test_vmap_caching(self):
+    # https://github.com/google/jax/issues/7621
+
+    f = lambda x: jnp.square(x).mean()
+    jf = jax.jit(f)
+    x = jax.random.uniform(jax.random.PRNGKey(0), shape=(8, 4))
+
+    with jtu.count_jit_and_pmap_compiles() as count:  # noqa: F841
+      jax.hessian(jf)(x).block_until_ready()
+      jax.hessian(jf)(x).block_until_ready()
+      jax.hessian(jf)(x).block_until_ready()
+
+    self.assertEqual(count[0], 2)
```

Comment:
I should probably have left a comment about this: the reason why the expected value is 2 is that this is a grad-of-jit type situation (really hessian-of-jit), so we have one compiled computation for the forward pass and one for the backward pass. If this were jit-of-grad we'd only get one compilation.